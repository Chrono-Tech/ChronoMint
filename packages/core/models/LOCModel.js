/**
 * Copyright 2017â€“2018, LaborX PTY
 * Licensed under the AGPL Version 3 license.
 */

import { I18n } from '@chronobank/core-dependencies/i18n'
import BigNumber from 'bignumber.js'
import moment from 'moment'

import { abstractFetchingModel } from './AbstractFetchingModel'
import Amount from './Amount'
import { dateFormatOptions } from './constants'
import TokenModel from './tokens/TokenModel'

export const THE_90_DAYS = 90 * 24 * 60 * 60 * 1000

export const STATUS_MAINTENANCE = 0
export const STATUS_ACTIVE = 1
export const STATUS_SUSPENDED = 2
export const STATUS_BANKRUPT = 3
export const STATUS_INACTIVE = 4

const statusesMeta = {
  [ STATUS_MAINTENANCE ]: {
    token: 'locs.status.maintenance',
    styleName: 'maintenance',
  },
  [ STATUS_ACTIVE ]: {
    token: 'locs.status.active',
    styleName: 'active',
  },
  [ STATUS_SUSPENDED ]: {
    token: 'locs.status.suspended',
    styleName: 'suspended',
  },
  [ STATUS_BANKRUPT ]: {
    token: 'locs.status.bankrupt',
    styleName: 'bankrupt',
  },
  [ STATUS_INACTIVE ]: {
    token: 'locs.status.inactive',
    styleName: 'inactive',
  },
}

class LOCModel extends abstractFetchingModel({
  name: '',
  oldName: '', // for update logic
  website: '',
  issued: new Amount(0),
  issueLimit: new Amount(0),
  publishedHash: '',
  expDate: Date.now() + THE_90_DAYS,
  createDate: Date.now(),
  status: 0,
  securityPercentage: 0,
  isNew: true,
  symbol: null,
}) {
  name (value) {
    return this._getSet('name', value)
  }

  website () {
    return this.get('website')
  }

  oldName (value) {
    return this._getSet('oldName', value)
  }

  issueLimit () {
    return this.get('issueLimit')
  }

  issued (value: BigNumber): BigNumber {
    return this._getSet('issued', value)
  }

  expDate () {
    return this.get('expDate')
  }

  expDateString () {
    return new Date(this.expDate()).toLocaleDateString('en-us', dateFormatOptions)
  }

  createDate () {
    return this.get('createDate')
  }

  createDateString () {
    return new Date(this.createDate()).toLocaleDateString('en-us', dateFormatOptions)
  }

  daysLeft () {
    return this.isActive() ? moment(this.expDate()).diff(Date.now(), 'days') : 0
  }

  status () {
    return this.isNotExpired() ? this.get('status') : STATUS_INACTIVE
  }

  statusString (status) {
    const statusId = status === undefined ? this.status() : status
    return I18n.t(statusesMeta[ statusId ].token)
  }

  statusStyle () {
    return statusesMeta[ this.status() ].styleName
  }

  symbol (value) {
    return this._getSet('symbol', value)
  }

  currency () {
    return this.symbol()
  }

  publishedHash () {
    return this.get('publishedHash')
  }

  isNew () {
    return this.get('isNew')
  }

  isNotExpired () {
    return this.expDate() > Date.now()
  }

  isActive () {
    return this.isNotExpired() && this.get('status') === STATUS_ACTIVE
  }

  toFormJS (token: TokenModel) {
    return {
      name: this.name(),
      website: this.website(),
      expDate: new Date(this.get('expDate')),
      issueLimit: token.removeDecimals(this.issueLimit()),
      publishedHash: this.publishedHash(),
    }
  }
}

export default LOCModel
