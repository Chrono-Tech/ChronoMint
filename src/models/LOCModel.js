import {Record as record} from 'immutable'

class LOCModel extends record({
  address: null,
  hasConfirmed: null,
  locName: null,
  website: null,
  controller: null,
  issueLimit: 0,
  issued: 0,
  redeemed: 0,
  publishedHash: null,
  expDate: new Date().getTime() + 7776000000,
  status: 0,
  isSubmitting: false,
  isFetching: false,
  isIssuing: false,
  isRedeeming: false
}) {
  getAddress () {
    return this.get('address')
  }

  name () {
    return this.get('locName') ? this.get('locName') : this.get('address')
  }

  issueLimit () {
    return this.get('issueLimit')
  }

  issued () {
    return this.get('issued')
  }

  redeemed () {
    return this.get('redeemed')
  }

  expDate () {
    return this.get('expDate')
  }

  status () {
    return this.get('status')
  }

  isSubmitting () {
    return this.get('isSubmitting')
  }

  isFetching () {
    return this.get('isFetching')
  }

  isIssuing () {
    return this.get('isIssuing')
  }

  isRedeeming () {
    return this.get('isRedeeming')
  }
}

export default LOCModel
