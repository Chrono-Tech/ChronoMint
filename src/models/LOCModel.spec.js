import LOCModel, { THE_90_DAYS } from './LOCModel'
import contractManagerDAO from 'dao/ContractsManagerDAO'

describe('LOC model', () => {
  it('should construct and return data', async () => {
    const locManager = await contractManagerDAO.getLOCManagerDAO()

    let model = new LOCModel({
      name: 'name',
      oldName: 'oldName',
      website: 'www',
      issueLimit: 1000,
      issued: 10,
      redeemed: 5,
      status: 1,
      isNew: false,
      token: locManager.getDefaultToken()
    })

    expect(model.name()).toBe('name')
    expect(model.oldName()).toBe('oldName')
    expect(model.issueLimit()).toBe(1000)
    expect(model.issued()).toBe(10)
    expect(Math.floor((model.expDate() - model.createDate())/1000)).toEqual(Math.floor(THE_90_DAYS/1000))
    expect(model.daysLeft()).toBe(90 - 1)
    expect(model.status()).toBe(1)
    expect(model.currency()).toBe('LHT')
    expect(model.isNew()).toBe(false)

    model = model.isPending(true)
    expect(model.isPending()).toBe(true)
    expect(model.isFailed()).toBe(false)

    model = model.isFailed(true)
    expect(model.isPending()).toBe(false)
    expect(model.isFailed()).toBe(true)
  })
})
