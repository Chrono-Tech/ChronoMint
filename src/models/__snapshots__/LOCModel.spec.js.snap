// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LOC model should construct and return data 1`] = `
{
  "createDate": "222222222222",
  "expDate": "11111111111",
  "id": null,
  "isFailed": false,
  "isFetched": false,
  "isFetching": false,
  "isInited": false,
  "isNew": false,
  "isPending": false,
  "isSelected": false,
  "issueLimit": 1000,
  "issued": 10,
  "name": "name",
  "oldName": "oldName",
  "publishedHash": "",
  "securityPercentage": 0,
  "status": 1,
  "symbol": null,
  "timestamp": null,
  "transactionHash": null,
  "website": "www"
}
`;
